specs/daito/Unique Durable Identifier for Volumes.md

```diff
@@ -0,0 +1,147 @@
+
+## Motivation
+Host platforms like VMware uses generic  algorithm  to generate a unique id for Virtual Disks. Management modules who manage end to end path (plugins or SRMs) can use this identifier to associate the relation . A common method used by Host Platforms is a  SCSI Inquiry command to the VPD page to retrieve   some "metadata" of  target LUN. This id will  persist between restarts , attaching to a different machine or a different SCSI port of the same machine. There are different standards followed in industry for unique identifier like NAA, EUI FC_WWN etc.
+
+Having this identifier details in volume spec would help these platforms to associate vDisk to Volume in OpenSDS.
+
+
+### Goals
+
+* Define an Identifier structure as used standard storage APIs like  swordfish.
+* Define modified  Volumespec with Identifier.
+* Include  Identifier in  API response.
+
+### Non-Goals
+
+* How driver assign identifier is not covered in this design doc.
+
+## Design Details
+
+### Identifier 
+This type describes any additional identifiers for a volume.
+
+
+| Property Name     | Type         | Description                              |
+|-------------------|--------------|------------------------------------------|
+| DurableName       | string       | This indicates the world wide, persistent name of the resource. |
+| DurableNameFormat | string(enum) | This represents the format of the DurableName property.<br>See DurableNameFormat in Property Details, below, for the possible values of this property. |
+
+#### Property details
+
+| string | Description                              |
+|--------|------------------------------------------|
+| EUI    | IEEE-defined 64-bit Extended Unique Identifier. |
+| FC_WWN | Fibre Channel World Wide Name.           |
+| iQN    | iSCSI Qualified Name.                    |
+| NAA    | Name Address Authority Format.           |
+| NQN    | NVMe Qualified Name.                     |
+| NSID   | NVM Namespace Identifier.                |
+| UUID   | Universally Unique Identifier.           |
+
+### Proposed VolumeSpec after adding Identfier
+
+```sh
+[
+  {
+    "id": "084bf71e-a102-11e7-88a8-e31fe6d52248",
+    "createdAt": {},
+    "updatedAt": {},
+    "tenantId": "string",
+    "userId": "string",
+    "name": "string",
+    "multiAttach": true,
+    "description": "string",
+    "size": 2,
+    "availabilityZone": "string",
+    "status": "string",
+    "profileId": "string",
+    "poolId": "string",
+    "snapshotId": "string",
+    "groupId": "string",
+    "snapshotFromCloud": true,
+    "replicationId": "string",
+    "replicationDriverData": {
+      "key1": "value1",
+      "key2": "value2"
+    },
+    "AttachStatus": "",
+    "identifier": {
+         "durableName": "value",
+        "durableNameFormat": "valueFormat"
+        }
+    "metadata": {
```

Comment:
if the volume contains more than two identifiers, how to express it?

Reply:
The identifierr will go to the VPD page of the device. Oly one identifier type will be there and it will be one of the type described above.

Reply:
Good work @NajmudheenCT
 I have few minor questions.

- Is the VolumeSpec versioned?  
   In the evolution of OpenSDS a future VolumeSpec will probably require a design for new features.

- Does `snapshotId` have a relationship with `replicationId`?  
   Should `replicationId` be null if `snapshotId` is null?
   
At a data management level, I like to consider a volume as being a fragment of large dataset so a snapshot would also belong to a parent dataset.  If the term `ReplicationId` is the parent association for set of volumes, is there a corresponding `BackupReplicationId` to associate a set of snapshotsId volumes to a parent.

Sorry **if my questions are outside the scope** - the PR scope is volume - I'm just having a first looke at this, and trying to picture the broader use case.

Reply:
Agree the 1st point that versioned API or model shall be included. 

For second question, IMO, there is no relationship between `snapshotId` and `relationshipId`. `snapshotId` means when a volume created from an existed snapshot. `relationshipId` is used in disaster recovery scenario when user wants to configure data replication between source volumes and target volumes.

Reply:
Thanks. Agreed.